### PR TITLE
SI-8092 More verify for f-interpolator

### PR DESCRIPTION
**Pull request by som-snytt**
Attempt to verify the nooks and crannies of the format string.

Allows all syntax in the javadoc, including arg indexes. If the
specifier after an arg has an index that doesn't refer to the arg,
a warning is issued and the missing `%s` is prepended (just as
for a part with a leading `%n`).

Other enhancements include detecting that a `Formattable` wasn't
supplied to `%#s`.

Error messages attempt to be pithy but descriptive.

### CODE REVIEW

**Comment by adriaanm** [comment]
Review by @densh please.

File: src/compiler/scala/tools/reflect/MacroImplementations.scala [hunk 1]
```diff
@@ -1,171 +1,296 @@
...
+      val fstring = bldr.toString
+      val expr =
+        Apply(
+          Select(
+            Literal(Constant(fstring)),
+            TermName("format")),
+          List(ids: _* )
+        )
+      Block(evals.toList, atPos(origApplyPos.focus)(expr)) setPos origApplyPos.makeTransparent
+      //q"..evals, ${Literal(Constant(fstring))}.format(..$ids)"
```
Comment (densh) [inline src/compiler/scala/tools/reflect/MacroImplementations.scala]: `q"..$evals; $fstring.format(..$ids)"`

File: src/compiler/scala/tools/reflect/MacroImplementations.scala [hunk 2]
```diff
@@ -1,171 +1,296 @@
...
+      variants find (arg.tpe <:< _)
+    // None if conforming to none
+    def checkType1(arg: Tree, variants: Type*): Option[Type] =
+      checkType0(arg, variants: _*) orElse (
+        variants find (c.inferImplicitView(arg, arg.tpe, _) != EmptyTree)
       )
-    }
+    // require variants.nonEmpty
+    def checkType(arg: Tree, variants: Type*): Option[Type] =
+      checkType1(arg, variants: _*) orElse Some(variants(0))
```
Comment (densh) [inline src/compiler/scala/tools/reflect/MacroImplementations.scala]: checkType functions could use a better name. Maybe:

```
checkType0 --> findSubtype
checkType1 --> findConforming
checkType  --> findConformingOrFirst
```

File: src/compiler/scala/tools/reflect/MacroImplementations.scala [hunk 3]
```diff
@@ -1,171 +1,296 @@
...
+   *  is inserted.
+   *
+   *  In any other position, the only permissible conversions are
+   *  the literals (%% and %n) or an index reference (%1$ or %<).
+   *
+   *  A conversion specifier has the form:
+   *
+   *  [index$][flags][width][.precision]conversion
+   */
+  def macro_StringInterpolation_f(parts: List[Tree], args: List[Tree], origApplyPos: c.universe.Position): Tree = {
```
Comment (densh) [inline src/compiler/scala/tools/reflect/MacroImplementations.scala]: Is it just me or this code has become a bit too nested?
I'd personally love to see something like:
1. rename `MacroImplementations` into `FInterpolator` (there are no more implementations in there)
2. make `def expandFInterpolator` an entry point without arguments
3. Move parsing of the macro prefix from `FastTrack` to top-level of the `FInterpolator`

This way you can flatten most of current `def macro_StringInterpolation_f` into the `FInterpolator` class and `expandFInterpolator` will be a clean entry point to that with minimal high-level code.  This scheme is very similar to what I have in quasiquotes and I think it's a win-win.

Comment (densh) [inline src/compiler/scala/tools/reflect/MacroImplementations.scala]: /cc @xeno-by

File: src/compiler/scala/tools/reflect/MacroImplementations.scala [hunk 4]
```diff
@@ -1,171 +1,296 @@
...
-          idx = 1
-        } else {
-          bldr append "%s"
-          defval(arg, AnyTpe)
+          val at = if (dk.op.isEmpty) Spec else CC
+          dk.errorAt(at, msg)
+        }
+        def cv(cc: Char) = cc match {
+          case 'b' | 'B' | 'h' | 'H' | 's' | 'S' =>
+            new Conversion(m, p) with GeneralXn
```
Comment (densh) [inline src/compiler/scala/tools/reflect/MacroImplementations.scala]: Why not `class FoobarXn(m, p) extends Conversion(m, p)` and `new FoobarXn(m, p)`? Current representation will create twice as much classes in the bytecode.

File: src/compiler/scala/tools/reflect/MacroImplementations.scala (hunk 3 above)
Comment (som-snytt) [inline src/compiler/scala/tools/reflect/MacroImplementations.scala]: I agree.  My focus was entirely on getting close to correctness.

I've only just discovered bundles. An amazing convenience.

File: src/compiler/scala/tools/reflect/MacroImplementations.scala [hunk 5]
```diff
@@ -1,171 +1,296 @@
...
-        "conversions must follow a splice; use %% for literal %, %n for newline")
-
-      // STEP 1: handle argument conversion
-      // 1) "...${smth}" => okay, equivalent to "...${smth}%s"
-      // 2) "...${smth}blahblah" => okay, equivalent to "...${smth}%sblahblah"
-      // 3) "...${smth}%" => error
-      // 4) "...${smth}%n" => okay, equivalent to "...${smth}%s%n"
-      // 5) "...${smth}%%" => okay, equivalent to "...${smth}%s%%"
-      // 6) "...${smth}[%legalJavaConversion]" => okay, according to http://docs.oracle.com/javase/1.5.0/docs/api/java/util/Formatter.html
-      // 7) "...${smth}[%illegalJavaConversion]" => error
```
Comment (som-snytt) [inline src/compiler/scala/tools/reflect/MacroImplementations.scala:107]: Maybe restore the comment in some form.

**Comment by som-snytt** [comment]
Took all of @densh 's good advice.  Many thanks for taking the time.  It's non-trivial LOC.

I was also looking at quasiquotes and parser, it's all pretty neat reading. So I esp appreciate time away from that for this.

File: src/compiler/scala/tools/reflect/FormatInterpolator.scala [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+  import c.universe.{ Match => _, _ }
+  import definitions._
+
+  @inline private def truly(body: => Unit): Boolean = { body ; true }
+  @inline private def falsely(body: => Unit): Boolean = { body ; false }
+
+  private def fail(msg: String) = c.abort(c.enclosingPosition, msg)
+
+  def interpolate: Tree = c.macroApplication match {
+    case q"$_(..$parts).f(..$args)" =>
```
Comment (densh) [inline src/compiler/scala/tools/reflect/FormatInterpolator.scala]: Unfortunately we can't use quasi-quotes in the 2.11 compiler code. Use treeInfo.Applied instead.

File: src/compiler/scala/tools/reflect/FormatInterpolator.scala [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        val (p, msg) =
+          if (parts.length == 0) (c.prefix.tree.pos, "there are no parts")
+          else if (args.length + 1 < parts.length)
+            (if (args.isEmpty) c.enclosingPosition else args.last.pos, because("few"))
+          else (args(parts.length-1).pos, because("many"))
+        c.abort(p, msg)
+      }
+      if (badlyInvoked) c.macroApplication else interpolated(parts, args)
+    case other =>
+      fail(s"Unexpected application ${showRaw(other)}")
```
Comment (densh) [inline src/compiler/scala/tools/reflect/FormatInterpolator.scala]: global.abort is probably more appropriate in situations like that.

**Comment by som-snytt** [comment]
Push without crazyquotes. They're totally crazy!

**Comment by som-snytt** [comment]
Pushed squashed without the (legacy) varargs args for `Apply`.  It took me a while to chase down what `Apply` is and which I want to invoke.

In the "refactor" commit, I changed `contexts.Context` to `runtime.Context` to look like the qq code.  I would have done a bit more to homogenize if I had time.

**Comment by densh** [comment]
LGTM

**Pull request merged**